modules/controller_pool/controller-primary.tpl

```diff
@@ -5,35 +5,40 @@ export WORKLOADS=$(echo ${workloads})
 mkdir $HOME/kube
 
 function load_workloads() {
-  echo "{"| tee -a $HOME/workloads.json ; for w in $WORKLOADS; do \ 
+  echo "{"| tee -a $HOME/workloads.json ; for w in $WORKLOADS; do \
   echo $w | sed 's| |\n|'g | awk '{sub(/:/,"\":\"")}1' | sed 's/.*/"&",/' | tee -a $HOME/workloads.json; \
   done ; echo "\"applied_at\":\"$(date +%F:%H:%m:%S)\"" | tee -a $HOME/workloads.json \
   ; echo "}" | tee -a $HOME/workloads.json
 }
 
 function install_containerd() {
-cat <<EOF > /etc/modules-load.d/containerd.conf
+  cat <<EOF > /etc/modules-load.d/containerd.conf
 overlay
 br_netfilter
 EOF
- modprobe overlay
- modprobe br_netfilter
- echo "Installing Containerd..."
- apt-get update
- apt-get install -y ca-certificates socat ebtables apt-transport-https cloud-utils prips containerd jq python3
+  modprobe overlay
+  modprobe br_netfilter
+  echo "Installing Containerd..."
+  apt-get update && apt-get install -y gnupg2 software-properties-common apt-transport-https ca-certificates socat ebtables cloud-utils prips jq python3
+  curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -
+  echo "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" > /etc/apt/sources.list.d/docker.list
+  apt-get update && apt-get install -y containerd.io
+  # Configure containerd
+  mkdir -p /etc/containerd
+  containerd config default | sudo tee /etc/containerd/config.toml >/dev/null 2>&1
+  sed -i 's/SystemdCgroup \= false/SystemdCgroup \= true/g' /etc/containerd/config.toml
 }
 
 function enable_containerd() {
- systemctl daemon-reload
- systemctl enable containerd
- systemctl start containerd
+  systemctl daemon-reload
+  systemctl restart containerd
+  systemctl enable containerd
 }
 
 function install_kube_tools {
- echo "Installing Kubeadm tools..." ;
- sed -ri '/\sswap\s/s/^#?/#/' /etc/fstab
```

Comment:
`sed -i.bak -E -e` might be a way to keep the `\s` patterns (which should be more versatile and forgiving to minor changes in the stock config file between OS versions).